### ISSUE TITLE
Download update improvements

### ISSUE THREAD

**Issue by klokan** [osm2vectortiles/osm2vectortiles#379]
Issues reported:
- MD5 is not correct for planet.mbtiles, after/during update
- Update of the planet is not atomic (link on the website points to a file which is being uploaded) so people downloading in the same moment never get a correct file

```sh
# md5sum planet.mbtiles
9821f7be1100f730165c6b394327ae46  planet.mbtiles
```

Report:
On Fri, Jul 15, 2016 at 7:28 AM, <ferns@backbonesystems.nl> wrote:
> Hi, Thanks for the wonderful OSM planet vector tile download you provide. Unfortunately, it seems the planet md5 on the download is not updated and it looks like the planet file download is not 'stable': I downloaded the planet file two times and I got different files with different checksums and the data in the western part of the Netherlands (around Amsterdam) in the deeper levers (around level 10) is not available...maybe because of corrupt data? I'm downloading the planet file now for a third time because it was updated (Thursday, 14 July 2016 at 18:57 UTC...maybe as I was downloading it?). Maybe you can look into a solution which makes the download more stable? But again, osm2vectortiles is a great site and it is perfect to set up vector tiles quickly. Great job! Cheers, Ferns

**Comment by brianshaler** [comment]
For what it's worth, I downloaded planet.mbtiles that day, am experiencing the same issue described in the email, 404s in northern Europe starting at zoom level >10. Interestingly, my checksum matches above ('9821f7be1100f730165c6b394327ae46').

**Comment by brianshaler** [comment]
Waiting on md5sum on the new download, but what I downloaded last night appears to the same as what I downloaded on the 14th (both files show same number of bytes and date modified matches upload date)

    57184074752 Jul 14 14:57 planet.mbtiles

**Comment by klokan** [comment]
@brianshaler Could you please create a new ticket where you mentioned the exact ZXY tiles which are missing in the dump?

**Comment by lukasmartinelli** [comment]
> MD5 is not correct for planet.mbtiles, after/during update

Yeah always need to update the md5 in the future.

> Update of the planet is not atomic (link on the website points to a file which is being uploaded) so people downloading in the same moment never get a correct file

Preserving old planets cost money but I think it is still worth the tradeoff.
Keep old versions of the planet around (but not of the extracts).

So process for the future (sadly already didn't follow this when I was reuploading the missing 500 z8 pyramids from #380 before):
 
1. Always give a distinct name of the planet file (add date?).
2. Calc md5 - add to web page
3. Regenerate extracts

**Comment by manuelroth** [comment]
### Planet filename proposal

**schema**: planet-{date}-{md5-hash}.mbtiles
**example**: planet-270716-a426f297c4936f97f2a14fc5a386e3dd.mbtiles

This way a user which has downloaded the planet file knows how current the planetfile is (date) and directly has the md5 hash to verify the file. Secondly, we don't have to update the md5 hash on the website on every upload anymore, because we can extract the hash programmatically based on the filename 🎉 

Is there a convention around using minus (-) or underline (_) as filename separator?- Thinking about harmonizing the filenames of the city and country extracts. Some of them have mixed filenames like the following: `v2.0/extracts/mountain-view_california.mbtiles`. I think we should decide on one.

@lukasmartinelli what do you think about this?- The [create-extract.py](https://github.com/osm2vectortiles/osm2vectortiles/blob/master/src/create-extracts/create_extracts.py) script would probably need to be modified to take the planet filename as argument. Couldn't find the code which uploads the planet file and planet zoom levels, is this done manually right now?

**Comment by lukasmartinelli** [comment]
> This way a user which has downloaded the planet file knows how current the planetfile is (date) and directly has the md5 hash to verify the file. Secondly, we don't have to update the md5 hash on the website on every upload anymore, because we can extract the hash programmatically based on the filename 🎉

Good idea for the planet. Will start using this.

> Is there a convention around using minus (-) or underline (_) as filename separator?- Thinking about harmonizing the filenames of the city and country extracts. Some of them use both (minus and underline) in their filename like the following: v2.0/extracts/mountain-view_california.mbtiles. I think we should decide on one.

`-` means space (`mountain-view` -> `Mountain View`).
`_` is the separator that separates the city from the region/country (originating from Metro extracts)


> @lukasmartinelli what do you think about this?- The create-extract.py script would probably need to be modified to take the planet filename as argument. So that it is possible to assign a different filename on each upload of the planet file.

I usually rename the file before doing it but adding a env var is easy.

>  Couldn't find the code which uploads the planet file and planet zoom levels, is this done manually right now?

Only planet is uploaded manually. Low zoom level extracts are automated (https://github.com/osm2vectortiles/osm2vectortiles/blob/master/src/create-extracts/create-extracts.sh).

**Comment by manuelroth** [comment]
> Good idea for the planet. Will start using this.

Cool.

> `-` means space (`mountain-view` -> `Mountain View`).
`_` is the separator that separates the city from the region/country (originating from Metro extracts)

Alright, now I understand. We will let it how it is right now. 

> I usually rename the file before doing it but adding a env var is easy.

So no code changes need to be made, great.

**Comment by ImreSamu** [comment]
Lot of improvements :)  

My question about the new download schemas :  
* in the future ...  I would like to write a daily cron script  for detecting the new (Planet/Hungary/Budapest)  [extract exists](http://osm2vectortiles.org/downloads/)  or not
 * and if  the new version is exists -  I would like to automatically download ,  check the md5 and refresh my local copy. 

What is the suggested method to implement this? This is a typical question we need to add to the FAQ.


other ..

We have 2 dates/timestamps 
* osm planet last modifications timestamp ( this is sometimes important for debugging, because the osm database is changing ,   for example  XXX POI is changed before or after ????  )  
   it is like "OSM Data from: 2016-07-27 00:58 UTC"
* mbtiles generating date/timestamp ( =  Last modified )  when the generating process finished, so this is a technical info - but also important for debugging

Is it possible to add both timestamp somewhere ?   
* now only - I can see this on the download page ...  
 ["Full Planet and extracts were last updated on Tuesday, 26 July 2016 at 19:24 UTC"](http://osm2vectortiles.org/downloads/)

**Comment by lukasmartinelli** [comment]
@ImreSamu The definition would be that when the planet file changes all extracts change as well. Cron script could actually detect the update date from the S3 bucket but that is not perfect.

I now use the new naming schema (and date and md5sum are derived from it) on http://osm2vectortiles.org/downloads/.

> MD5 is not correct for planet.mbtiles, after/during update
Update of the planet is not atomic (link on the website points to a file which is being uploaded) so people downloading in the same moment never get a correct file

This is now solved since the link is atomic now.

Links to the planet and low zoom level extracts.
The low zoom level extracts always contain the md5sum of the planet file they were derived form.

```
s3://osm2vectortiles-downloads/v2.0/planet_2016-06-20_7088ce06a738dcb3104c769adc11ac2c_z0-z14.mbtiles
s3://osm2vectortiles-downloads/v2.0/planet_2016-06-20_7088ce06a738dcb3104c769adc11ac2c_z0-z5.mbtiles
s3://osm2vectortiles-downloads/v2.0/planet_2016-06-20_7088ce06a738dcb3104c769adc11ac2c_z0-z8.mbtiles
```

I also deleted the previous planet file that has no md5sum in it (I know this now breaks the link for people but otherwise they would always get the old version).


> We have 2 dates/timestamps
> 
> osm planet last modifications timestamp ( this is sometimes important for debugging, because the osm database is changing , for example XXX POI is changed before or after ???? )
> it is like "OSM Data from: 2016-07-27 00:58 UTC"
> mbtiles generating date/timestamp ( = Last modified ) when the generating process finished, so this is a technical info - but also important for debugging

@ImreSamu We should open a separate issue for that. I think adding metadata to the MBTiles is a good solution (we could even embed the information from the `osm_timestamp` table)
